src/api/process.c

```diff
@@ -63,60 +84,210 @@ typedef enum {
+  SDL_LockMutex(list->mutex);
+  // wait until list is empty
+  while (list->size)
+    SDL_CondWait(list->work_done, list->mutex);
+  // tell the worker to stop
+  list->stop = true;
+  SDL_CondSignal(list->has_work);
+  SDL_UnlockMutex(list->mutex);
+}
+
+
+static void process_handle_close(process_handle *handle) {
+#ifdef _WIN32
+  if (*handle) {
+    CloseHandle(*handle);
+    *handle = NULL;
+  }
 #endif
+  (void) 0;
+}
+
+
+static bool process_handle_is_running(process_handle handle, int *status) {
+#ifdef _WIN32
+  DWORD s;
+  if (GetExitCodeProcess(handle, &s) && s != STILL_ACTIVE) {
+    if (status != NULL)
+      *status = s;
+    return false;
+  }
+#else
+  int s;
+  if (waitpid(handle, &s, WNOHANG) != 0) {
+    if (status != NULL)
+      *status = WEXITSTATUS(s);
+    return false;
+  }
+#endif
+  return true;
+}
+
+
+static bool process_handle_signal(process_handle handle, signal_e sig) {
+#if _WIN32
+  switch(sig) {
+    case SIGNAL_TERM: return GenerateConsoleCtrlEvent(CTRL_BREAK_EVENT, GetProcessId(handle));
+    case SIGNAL_KILL: return TerminateProcess(handle, -1);
+    case SIGNAL_INTERRUPT: return DebugBreakProcess(handle);
+  }
+#else
+  switch (sig) {
+    case SIGNAL_TERM: return kill(-handle, SIGTERM) == 0; break;
+    case SIGNAL_KILL: return kill(-handle, SIGKILL) == 0; break;
+    case SIGNAL_INTERRUPT: return kill(-handle, SIGINT) == 0; break;
+  }
+#endif
+  return false;
+}
+
+
+static int kill_list_worker(void *ud) {
+  process_kill_list_t *list = (process_kill_list_t *) ud;
+  process_kill_t *current_task;
+  int i, size;
+
+  while (true) {
+    SDL_LockMutex(list->mutex);
+
+    // wait until we have work to do
+    while (!list->head && !list->stop)
+      SDL_CondWait(list->has_work, list->mutex); // LOCK MUTEX
+
+    if (list->stop) break;
+
+    for (i = 0, size = list->size; i < size; i++) {
+      current_task = list->head;
+      if ((SDL_GetTicks() - current_task->start_time) < PROCESS_TERM_DELAY)
+        break;
+      kill_list_pop(list);
+      if (process_handle_is_running(current_task->handle, NULL)) {
+        if (current_task->tries < PROCESS_TERM_TRIES)
+          process_handle_signal(current_task->handle, SIGNAL_TERM);
+        else if (current_task->tries == PROCESS_TERM_TRIES)
+          process_handle_signal(current_task->handle, SIGNAL_KILL);
+        else
+          goto free_task;
+
+        // add the task back into the queue
+        current_task->tries++;
+        current_task->start_time = SDL_GetTicks();
+        kill_list_push(list, current_task);
+      } else {
+        free_task:
+        process_handle_close(&current_task->handle);
+        free(current_task);
+      }
+    }
+    if (i > 0)
+      SDL_CondSignal(list->work_done); // only signal when we did something
+    // if the queue is empty, we can immediately wait for another task or to quit.
+    if (list->size) {
+      SDL_UnlockMutex(list->mutex);
+      SDL_Delay(PROCESS_TERM_DELAY);
```

Comment:
Could we delay the amount of time needed for the task to timeout?
This way if a task only needed 10ms to timeout, we wouldn't waste 40ms waiting.